File: port_scanner.go
// port_scanner.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

type ScanResult struct {
	Port   int    `json:"port"`
	State  string `json:"state"`
	Banner string `json:"banner"`
}

type Scanner struct {
	host    string
	ip      net.IP
	ports   []int
	timeout time.Duration
	threads int
	verbose bool
	color   bool
	results []ScanResult
	mu      sync.Mutex
}

func NewScanner(host string, portStr string, timeout float64, threads int, verbose, color bool) (*Scanner, error) {
	ip, err := net.ResolveIPAddr("ip", host)
	if err != nil {
		return nil, err
	}
	ports, err := parsePorts(portStr)
	if err != nil {
		return nil, err
	}
	return &Scanner{
		host:    host,
		ip:      ip.IP,
		ports:   ports,
		timeout: time.Duration(timeout * float64(time.Second)),
		threads: threads,
		verbose: verbose,
		color:   color,
	}, nil
}

func parsePorts(portStr string) ([]int, error) {
	var ports []int
	for _, part := range strings.Split(portStr, ",") {
		if strings.Contains(part, "-") {
			r := strings.Split(part, "-")
			start, err := strconv.Atoi(r[0])
			if err != nil {
				return nil, err
			}
			end, err := strconv.Atoi(r[1])
			if err != nil {
				return nil, err
			}
			for p := start; p <= end; p++ {
				ports = append(ports, p)
			}
		} else {
			p, err := strconv.Atoi(part)
			if err != nil {
				return nil, err
			}
			ports = append(ports, p)
		}
	}
	return ports, nil
}

func (s *Scanner) scanPort(port int) ScanResult {
	addr := net.TCPAddr{IP: s.ip, Port: port}
	conn, err := net.DialTimeout("tcp", addr.String(), s.timeout)
	if err != nil {
		if strings.Contains(err.Error(), "refused") {
			return ScanResult{Port: port, State: "closed", Banner: ""}
		}
		if strings.Contains(err.Error(), "timeout") {
			return ScanResult{Port: port, State: "filtered", Banner: ""}
		}
		return ScanResult{Port: port, State: "error", Banner: err.Error()}
	}
	defer conn.Close()
	return ScanResult{Port: port, State: "open", Banner: ""}
}

func (s *Scanner) worker(ports <-chan int, wg *sync.WaitGroup) {
	defer wg.Done()
	for port := range ports {
		res := s.scanPort(port)
		s.mu.Lock()
		s.results = append(s.results, res)
		if s.verbose {
			s.printResult(res)
		}
		s.mu.Unlock()
	}
}

func (s *Scanner) printResult(res ScanResult) {
	stateColor := ""
	switch res.State {
	case "open":
		stateColor = "\033[32m"
	case "closed":
		stateColor = "\033[31m"
	case "filtered":
		stateColor = "\033[33m"
	default:
		stateColor = "\033[37m"
	}
	if s.color {
		fmt.Printf("%sPort %d/tcp  %s\033[0m  %s\n", stateColor, res.Port, res.State, res.Banner)
	} else {
		fmt.Printf("Port %d/tcp  %s  %s\n", res.Port, res.State, res.Banner)
	}
}

func (s *Scanner) Scan() {
	fmt.Printf("Scanning %s (%s)...\n", s.host, s.ip.String())
	startTime := time.Now()
	var wg sync.WaitGroup
	portsChan := make(chan int, len(s.ports))
	for i := 0; i < s.threads; i++ {
		wg.Add(1)
		go s.worker(portsChan, &wg)
	}
	for _, p := range s.ports {
		portsChan <- p
	}
	close(portsChan)
	wg.Wait()
	elapsed := time.Since(startTime).Seconds()
	openCount := 0
	for _, r := range s.results {
		if r.State == "open" {
			openCount++
		}
	}
	fmt.Printf("\nScan completed in %.2fs. Found %d open ports.\n", elapsed, openCount)
}

func (s *Scanner) ExportJSON(filename string) error {
	data := struct {
		Host    string       `json:"host"`
		IP      string       `json:"ip"`
		Results []ScanResult `json:"results"`
	}{s.host, s.ip.String(), s.results}
	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, b, 0644)
}

func (s *Scanner) ExportCSV(filename string) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"port", "state", "banner"})
	for _, r := range s.results {
		w.Write([]string{strconv.Itoa(r.Port), r.State, r.Banner})
	}
	return nil
}

func main() {
	var (
		host    string
		ports   string
		timeout float64
		threads int
		jsonOut string
		csvOut  string
		verbose bool
		noColor bool
	)
	flag.StringVar(&host, "host", "", "Target host")
	flag.StringVar(&ports, "ports", "22,80,443,1-1024", "Ports to scan")
	flag.Float64Var(&timeout, "timeout", 2.0, "Timeout in seconds")
	flag.IntVar(&threads, "threads", 10, "Number of threads")
	flag.StringVar(&jsonOut, "json", "", "Export to JSON")
	flag.StringVar(&csvOut, "csv", "", "Export to CSV")
	flag.BoolVar(&verbose, "verbose", false, "Verbose output")
	flag.BoolVar(&noColor, "no-color", false, "Disable color")
	flag.Parse()

	if host == "" {
		fmt.Fprintln(os.Stderr, "Error: --host is required")
		os.Exit(1)
	}
	color := !noColor && isTerminal()
	scanner, err := NewScanner(host, ports, timeout, threads, verbose, color)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	scanner.Scan()
	if jsonOut != "" {
		if err := scanner.ExportJSON(jsonOut); err != nil {
			fmt.Fprintf(os.Stderr, "Export JSON error: %v\n", err)
		} else {
			fmt.Printf("Results exported to %s\n", jsonOut)
		}
	}
	if csvOut != "" {
		if err := scanner.ExportCSV(csvOut); err != nil {
			fmt.Fprintf(os.Stderr, "Export CSV error: %v\n", err)
		} else {
			fmt.Printf("Results exported to %s\n", csvOut)
		}
	}
}

func isTerminal() bool {
	stat, _ := os.Stdout.Stat()
	return (stat.Mode() & os.ModeCharDevice) != 0
}

File: port_scanner.rs
// port_scanner.rs
use clap::{App, Arg};
use crossbeam::channel;
use std::net::{IpAddr, TcpStream, ToSocketAddrs};
use std::sync::{Arc, Mutex};
use std::time::Duration;
use serde::{Deserialize, Serialize};
use colored::*;

#[derive(Serialize, Deserialize, Debug, Clone)]
struct ScanResult {
    port: u16,
    state: String,
    banner: String,
}

struct Scanner {
    host: String,
    ip: IpAddr,
    ports: Vec<u16>,
    timeout: Duration,
    threads: usize,
    verbose: bool,
    color: bool,
    results: Arc<Mutex<Vec<ScanResult>>>,
}

impl Scanner {
    fn new(host: &str, port_str: &str, timeout: f64, threads: usize, verbose: bool, color: bool) -> Result<Self, Box<dyn std::error::Error>> {
        let ip = host.parse::<IpAddr>().or_else(|_| {
            let addrs = (host, 0).to_socket_addrs()?;
            let addr = addrs.into_iter().next().map(|s| s.ip()).ok_or("No IP")?;
            Ok(addr)
        })?;
        let ports = parse_ports(port_str)?;
        Ok(Scanner {
            host: host.to_string(),
            ip,
            ports,
            timeout: Duration::from_secs_f64(timeout),
            threads,
            verbose,
            color,
            results: Arc::new(Mutex::new(Vec::new())),
        })
    }

    fn scan_port(&self, port: u16) -> ScanResult {
        let addr = (self.ip, port);
        match TcpStream::connect_timeout(&addr, self.timeout) {
            Ok(_) => ScanResult { port, state: "open".to_string(), banner: "".to_string() },
            Err(e) => {
                if e.kind() == std::io::ErrorKind::ConnectionRefused {
                    ScanResult { port, state: "closed".to_string(), banner: "".to_string() }
                } else if e.kind() == std::io::ErrorKind::TimedOut {
                    ScanResult { port, state: "filtered".to_string(), banner: "".to_string() }
                } else {
                    ScanResult { port, state: "error".to_string(), banner: e.to_string() }
                }
            }
        }
    }

    fn print_result(&self, res: &ScanResult) {
        let state_colored = match res.state.as_str() {
            "open" => res.state.green(),
            "closed" => res.state.red(),
            "filtered" => res.state.yellow(),
            _ => res.state.normal(),
        };
        if self.color {
            println!("{}Port {}/tcp  {}  {}", state_colored, res.port, res.state, res.banner);
        } else {
            println!("Port {}/tcp  {}  {}", res.port, res.state, res.banner);
        }
    }

    fn scan(&self) {
        println!("Scanning {} ({})...", self.host, self.ip);
        let start_time = std::time::Instant::now();
        let (tx, rx) = channel::bounded::<u16>(self.ports.len());
        for &p in &self.ports {
            tx.send(p).unwrap();
        }
        drop(tx);
        let results = self.results.clone();
        let scanner = self;
        let handles: Vec<_> = (0..self.threads).map(|_| {
            let rx = rx.clone();
            let results = results.clone();
            std::thread::spawn(move || {
                for port in rx {
                    let res = scanner.scan_port(port);
                    {
                        let mut r = results.lock().unwrap();
                        r.push(res.clone());
                    }
                    if scanner.verbose {
                        scanner.print_result(&res);
                    }
                }
            })
        }).collect();
        for h in handles { h.join().unwrap(); }
        let elapsed = start_time.elapsed().as_secs_f64();
        let open_count = self.results.lock().unwrap().iter().filter(|r| r.state == "open").count();
        println!("\nScan completed in {:.2}s. Found {} open ports.", elapsed, open_count);
    }

    fn export_json(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let data = serde_json::json!({
            "host": self.host,
            "ip": self.ip.to_string(),
            "results": &*self.results.lock().unwrap(),
        });
        let json = serde_json::to_string_pretty(&data)?;
        std::fs::write(filename, json)?;
        Ok(())
    }

    fn export_csv(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut wtr = csv::Writer::from_path(filename)?;
        wtr.write_record(&["port", "state", "banner"])?;
        for r in self.results.lock().unwrap().iter() {
            wtr.write_record(&[r.port.to_string(), r.state.clone(), r.banner.clone()])?;
        }
        wtr.flush()?;
        Ok(())
    }
}

fn parse_ports(port_str: &str) -> Result<Vec<u16>, Box<dyn std::error::Error>> {
    let mut ports = Vec::new();
    for part in port_str.split(',') {
        if part.contains('-') {
            let range: Vec<&str> = part.split('-').collect();
            let start: u16 = range[0].parse()?;
            let end: u16 = range[1].parse()?;
            for p in start..=end {
                ports.push(p);
            }
        } else {
            ports.push(part.parse()?);
        }
    }
    Ok(ports)
}

fn main() {
    let matches = App::new("TCP Port Scanner")
        .arg(Arg::with_name("host").long("host").takes_value(true).required(true).help("Target host"))
        .arg(Arg::with_name("ports").long("ports").takes_value(true).default_value("22,80,443,1-1024").help("Ports to scan"))
        .arg(Arg::with_name("timeout").long("timeout").takes_value(true).default_value("2.0").help("Timeout in seconds"))
        .arg(Arg::with_name("threads").long("threads").takes_value(true).default_value("10").help("Number of threads"))
        .arg(Arg::with_name("json").long("json").takes_value(true).help("Export to JSON"))
        .arg(Arg::with_name("csv").long("csv").takes_value(true).help("Export to CSV"))
        .arg(Arg::with_name("verbose").long("verbose").help("Verbose output"))
        .arg(Arg::with_name("no-color").long("no-color").help("Disable color"))
        .get_matches();

    let host = matches.value_of("host").unwrap();
    let ports = matches.value_of("ports").unwrap();
    let timeout: f64 = matches.value_of("timeout").unwrap().parse().unwrap();
    let threads: usize = matches.value_of("threads").unwrap().parse().unwrap();
    let verbose = matches.is_present("verbose");
    let no_color = matches.is_present("no-color");
    let color = !no_color && atty::is(atty::Stream::Stdout);

    let scanner = Scanner::new(host, ports, timeout, threads, verbose, color)
        .expect("Failed to initialize scanner");
    scanner.scan();

    if let Some(json) = matches.value_of("json") {
        if let Err(e) = scanner.export_json(json) {
            eprintln!("Export JSON error: {}", e);
        } else {
            println!("Results exported to {}", json);
        }
    }
    if let Some(csv) = matches.value_of("csv") {
        if let Err(e) = scanner.export_csv(csv) {
            eprintln!("Export CSV error: {}", e);
        } else {
            println!("Results exported to {}", csv);
        }
    }
}
